### PR TITLE
copy mingpt + graph entire GPT + visualization + preliminary evolution code

### PR DESCRIPTION
**Pull request by PatrickNercessian**
Can ignore everything in mingpt directory, that's just copied from https://github.com/karpathy/minGPT/tree/master/mingpt

### CODE REVIEW

File: src/individual_graph_module.py
Comment (wmatray) [inline src/individual_graph_module.py:1]: I don't see any issues here but can we just double check that we still have the full graph functionality with the subclass

Comment (PatrickNercessian) [reply src/individual_graph_module.py:1]: Yeah I did - it seems like it, since the evolution is able to begin training (i let it run for about 30 seconds and went from a loss of 7-something to 3-something

**Review by wmatray — APPROVED** [review]

File: src/bpe.py
Comment (MortimerE) [inline src/bpe.py:1]: I accidentally reviewed the entire mingpt bpe implementation - 
This one appears to use substring frequency instead of recursive merging 
Is there a reason we deviate from the gpt-2 style encodings? i.e. is this better, and is there a literature source explaining why/how?

File: src/evaluate.py [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+    """
+    trainer = Trainer(individual.train_config, individual.graph_module, train_data_loader)
+    def batch_end_callback(trainer):
+        if trainer.iter_num % 100 == 0:
+            print(f"iter_dt {trainer.iter_dt * 1000:.2f}ms; iter {trainer.iter_num}: train loss {trainer.loss.item():.5f}")
+    trainer.set_callback('on_batch_end', batch_end_callback)
+    trainer.run()
+
+    # Calculate perplexity on the validation set
+    perplexity = calculate_perplexity(individual.graph_module, val_data_loader, device=device)
```
Comment (MortimerE) [inline src/evaluate.py:34]: More of a semantic question but is there a difference between using CE and perplexity here? p is just exp(ce) right

Comment (MortimerE) [reply src/evaluate.py:34]: I think overall I'd like to explore more complicated fitness functions but this implementation looks good

File: src/evolution.py
Comment (MortimerE) [inline src/evolution.py:1]: Assuming we'll revisit this when the functionality for the mutations is in place 

Based on our conversations so far would it make sense to have multiple mutation params for different mutation categories, then we could run ablations based on "gridsearching" through those?

File: src/individual.py
Comment (MortimerE) [inline src/individual.py:1]: What are we using the deepcopy for?

File: src/individual_graph_module.py [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+        We are separating out all parameters of the model into two buckets: those that will experience
+        weight decay for regularization and those that won't (biases, and layernorm/embedding weights).
+        We are then returning the PyTorch optimizer object.
+        """
+
+        # separate out all parameters to those that will and won't experience regularizing weight decay
+        decay = set()
+        no_decay = set()
+        whitelist_weight_modules = (torch.nn.Linear, )
+        blacklist_weight_modules = (torch.nn.LayerNorm, torch.nn.Embedding)
```
Comment (MortimerE) [inline src/individual_graph_module.py:19]: What kind of weights would we blacklist? Is this for like dropout regularization?

File: src/bpe.py
Comment (PatrickNercessian) [reply src/bpe.py:1]: Copied from Discord:

"Yeah I see that my BPE algorithm is not BPE at all lol, must have misunderstood BPE when I wrote it like 18 months ago"

File: src/bpe.py [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+If you be maid or no?
+
+MIRANDA:
+No wonder, sir;
+But certainly a maid.
+
+FERDINAND:
+My language! heavens!
+I am the best of them that speak this speech,
+Were I but where 'tis spoken."""))
```
[the remark this answers is not in the document]
Comment (PatrickNercessian) [reply src/bpe.py:128]: Yeah we can move things around down the line, I probably should have done it to begin with, but already merged

File: src/evaluate.py (hunk 1 above)
Comment (PatrickNercessian) [reply src/evaluate.py:34]: Yes perplexity is just exp(CE).

I believe the ranking would be the same as just using the CE loss, so it actually doesn't matter (if our selection process just takes the n most fit individuals).

File: src/evolution.py
Comment (PatrickNercessian) [reply src/evolution.py:1]: Yeah, in [my open PR](https://github.com/PatrickNercessian/evolving-dnn/pull/3), I've modified it to take a list of functions and their respective probability rates as parameters, so we can easily run different experiments. Did the same for crossover.

File: src/individual.py
Comment (PatrickNercessian) [reply src/individual.py:1]: We use it in https://github.com/PatrickNercessian/evolving-dnn/pull/3, to avoid persisting weight updates across generations, and to create new individuals during mutation/crossover.

**Review by MortimerE — COMMENTED** [review]
Nice BPE bro

File: src/evolution.py
Comment (MortimerE) [reply src/evolution.py:1]: Banger

File: src/individual.py
Comment (MortimerE) [reply src/individual.py:1]: Keeping it for creating new individuals makes sense, but to clarify are you saying we would not want the weights to persist into the next generation for the selected population? This would be to balance the population candidates so we aren't biasing the evolution towards more trained models, or screwing things up by training too far in one direction and then carrying those weights to a totally different architecture? 

I know some transfer learning techniques and finetuning in general basically depends on giving a model a pretrained set of weights to start on to do few-shot learning, maybe allowing persistent weights is could be useful in an ablation, but if I'm following correctly I think this makes sense to have as the primary/base implementation

File: src/bpe.py
Comment (MortimerE) [reply src/bpe.py:1]: wow, exposed

File: src/individual.py
Comment (PatrickNercessian) [reply src/individual.py:1]: > Keeping it for creating new individuals makes sense, but to clarify are you saying we would not want the weights to persist into the next generation for the selected population? This would be to balance the population candidates so we aren't biasing the evolution towards more trained models, or screwing things up by training too far in one direction and then carrying those weights to a totally different architecture?

Yeah exactly

> I know some transfer learning techniques and finetuning in general basically depends on giving a model a pretrained set of weights to start on to do few-shot learning, maybe allowing persistent weights is could be useful in an ablation, but if I'm following correctly I think this makes sense to have as the primary/base implementation

Agreed, Matray and I also have some discussion on this in the linked PR.

File: src/individual_graph_module.py (hunk 2 above)
Comment (MortimerE) [reply src/individual_graph_module.py:19]: Got it, I think we will end up using it for at least dropout but I just wasn't sure
